xcm/src/v3/multiasset.rs

```diff
@@ -0,0 +1,871 @@
+						(
+							MultiAsset { fun: Fungibility::Fungible(a_amount), id: a_id },
+							MultiAsset { fun: Fungibility::Fungible(b_amount), id: b_id },
+						) if a_id == b_id => MultiAsset {
+							id: a_id,
+							fun: Fungibility::Fungible(a_amount.saturating_add(b_amount)),
+						},
+						(
+							MultiAsset { fun: Fungibility::NonFungible(a_instance), id: a_id },
+							MultiAsset { fun: Fungibility::NonFungible(b_instance), id: b_id },
+						) if a_id == b_id && a_instance == b_instance =>
+							MultiAsset { fun: Fungibility::NonFungible(a_instance), id: a_id },
+						(to_push, to_remember) => {
+							res.push(to_push);
+							to_remember
+						},
+					}
+				});
+				res.push(last);
+			}
+		}
+		Self(res)
+	}
+}
+
+impl<T: Into<MultiAsset>> From<T> for MultiAssets {
+	fn from(x: T) -> Self {
+		Self(vec![x.into()])
+	}
+}
+
+impl MultiAssets {
+	/// A new (empty) value.
+	pub fn new() -> Self {
+		Self(Vec::new())
+	}
+
+	/// Create a new instance of `MultiAssets` from a `Vec<MultiAsset>` whose contents are sorted and
+	/// which contain no duplicates.
+	///
+	/// Returns `Ok` if the operation succeeds and `Err` if `r` is out of order or had duplicates. If you can't
+	/// guarantee that `r` is sorted and deduplicated, then use `From::<Vec<MultiAsset>>::from` which is infallible.
+	pub fn from_sorted_and_deduplicated(r: Vec<MultiAsset>) -> Result<Self, ()> {
+		if r.is_empty() {
+			return Ok(Self(Vec::new()))
+		}
+		r.iter().skip(1).try_fold(&r[0], |a, b| -> Result<&MultiAsset, ()> {
+			if a.id < b.id || a < b && (a.is_non_fungible(None) || b.is_non_fungible(None)) {
+				Ok(b)
+			} else {
+				Err(())
+			}
+		})?;
+		Ok(Self(r))
+	}
+
+	/// Create a new instance of `MultiAssets` from a `Vec<MultiAsset>` whose contents are sorted and
+	/// which contain no duplicates.
+	///
+	/// In release mode, this skips any checks to ensure that `r` is correct, making it a negligible-cost operation.
+	/// Generally though you should avoid using it unless you have a strict proof that `r` is valid.
+	#[cfg(test)]
+	pub fn from_sorted_and_deduplicated_skip_checks(r: Vec<MultiAsset>) -> Self {
+		Self::from_sorted_and_deduplicated(r).expect("Invalid input r is not sorted/deduped")
+	}
+	/// Create a new instance of `MultiAssets` from a `Vec<MultiAsset>` whose contents are sorted and
+	/// which contain no duplicates.
+	///
+	/// In release mode, this skips any checks to ensure that `r` is correct, making it a negligible-cost operation.
+	/// Generally though you should avoid using it unless you have a strict proof that `r` is valid.
+	///
+	/// In test mode, this checks anyway and panics on fail.
+	#[cfg(not(test))]
+	pub fn from_sorted_and_deduplicated_skip_checks(r: Vec<MultiAsset>) -> Self {
+		Self(r)
+	}
+
+	/// Add some asset onto the list, saturating. This is quite a laborious operation since it maintains the ordering.
+	pub fn push(&mut self, a: MultiAsset) {
+		if let Fungibility::Fungible(ref amount) = a.fun {
```

Comment:
Why does this not saturate non-fungibles like the `From<Vec<MultiAsset>> for MultiAssets` does?

Reply:
It would be unexpected and indicate a bug in the calling code for there to be two instances of the same non-fungible. However in the interests of consistency we probably should gracefully fail in the same way. @KiChjang could you do this?